File: android/src/main/java/org/ligi/passandroid/model/AndroidFileSystemPassStore.java
package org.ligi.passandroid.model;

import android.content.Context;
import org.ligi.passandroid.R;

import org.ligi.axt.AXT;
import org.ligi.passandroid.helper.DirectoryFileFilter;
import org.ligi.passandroid.reader.AppleStylePassReader;
import org.ligi.passandroid.reader.PassReader;
import org.ligi.tracedroid.logging.Log;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class AndroidFileSystemPassStore implements PassStore {

    private final Context context;
    private final String path;

    private List<FiledPass> passList = new ArrayList<>();
    private Pass actPass;
    private final PassClassifier passClassifier;

    public AndroidFileSystemPassStore(final Context context, final Settings settings) {
        this.context = context;
        path = settings.getPassesDir();

        refreshPassesList();
        final File classificationFile = new File(settings.getStateDir(), "classifier_state.json");
        passClassifier = new FileBackedPassClassifier(classificationFile, this, context.getString(R.string.topic_active));
    }

    @Override
    public void deleteCacheForId(String id) {
        getCacheFile(id).delete();
    }

    private File getCacheFile(String id) {
        return new File(getPathForID(id) + "/base_cache.obj");
    }

    @Override
    public List<FiledPass> getPassList() {
        return passList;
    }

    @Override
    public void preCachePassesList() {
        for (String key : getPassIDArray()) {
            getPassbookForId(key);
        }
    }

    @Override
    public void refreshPassesList() {

        final List<String> newIds = Arrays.asList(getPassIDArray());
        final List<String> oldIds = new ArrayList<>();
        final List<FiledPass> toRemove = new ArrayList<>();

        for (FiledPass pass : passList) {
            oldIds.add(pass.getId());
            if (!newIds.contains(pass.getId())) {
                toRemove.add(pass);
            }
        }

        for (String newId : newIds) {
            if (!oldIds.contains(newId)) {
                passList.add(getCachedPassOrLoad(newId));
            }
        }

        passList.removeAll(toRemove);
    }

    private String[] getPassIDArray() {
        return getPassesDirSafely().list(new DirectoryFileFilter());
    }


    private FiledPass getCachedPassOrLoad(String id) {
        final File cachedFile = getCacheFile(id);
        try {
            return AXT.at(cachedFile).loadToObject();
        } catch (Exception ignored) {
        }

        Log.i("Passbook cache miss");
        final String language = context.getResources().getConfiguration().locale.getLanguage();

        final FiledPass pass = readPass(id, language);

        AXT.at(cachedFile).writeObject(pass);
        return pass;
    }

    private FiledPass readPass(String id, String language) {
        if (new File(getPathForID(id), "data.json").exists()) {
            return PassReader.read(getPathForID(id));
        } else {
            return AppleStylePassReader.read(getPathForID(id), language);
        }
    }

    private File getPassesDirSafely() {
        final File passes_dir = new File(path);

        if (!passes_dir.exists()) {
            passes_dir.mkdirs();
        }

        return passes_dir;
    }

    @Override
    public Pass getPassbookForId(final String id) {
        for (Pass pass : passList) {
            if (pass.getId().equals(id)) {
                return pass;
            }
        }

        return getCachedPassOrLoad(id);
    }

    @Override
    public Pass getCurrentPass() {
        return actPass;
    }

    @Override
    public void setCurrentPass(final Pass pass) {
        actPass = pass;
    }

    @Override
    public PassClassifier getClassifier(Context context) {
        return passClassifier;
    }

    @Override
    public boolean deletePassWithId(final String id) {
        final boolean result = AXT.at(new File(getPathForID(id))).deleteRecursive();
        if (result) {
            refreshPassesList();
            passClassifier.processDataChange();
            passClassifier.notifyDataChange();
        }
        return result;
    }

    public String getPathForID(final String id) {
        return path + "/" + id;
    }

}

File: android/src/main/java/org/ligi/passandroid/model/AndroidFileSystemPassStore.kt
package org.ligi.passandroid.model

import android.content.Context
import com.squareup.moshi.JsonDataException
import com.squareup.moshi.Moshi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.channels.ConflatedBroadcastChannel
import kotlinx.coroutines.launch
import okio.buffer
import okio.sink
import okio.source
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import org.ligi.passandroid.BuildConfig
import org.ligi.passandroid.Tracker
import org.ligi.passandroid.model.pass.Pass
import org.ligi.passandroid.model.pass.PassImpl
import org.ligi.passandroid.reader.AppleStylePassReader
import org.ligi.passandroid.reader.PassReader
import java.io.File
import java.util.*

object PassStoreUpdateEvent

class AndroidFileSystemPassStore(
        private val context: Context,
        settings: Settings,
        private val moshi: Moshi
) : PassStore, KoinComponent {

    override val updateChannel = ConflatedBroadcastChannel<PassStoreUpdateEvent>()

    private val path: File = settings.getPassesDir()

    override val passMap = HashMap<String, Pass>()

    override var currentPass: Pass? = null

    private val tracker: Tracker by inject()

    override val classifier: PassClassifier by lazy {
        val classificationFile = File(settings.getStateDir(), "classifier_state.json")
        FileBackedPassClassifier(classificationFile, this, moshi)
    }

    override fun save(pass: Pass) {
        val jsonAdapter = moshi.adapter(PassImpl::class.java)

        val pathForID = getPathForID(pass.id)

        if (!pathForID.exists()) {
            pathForID.mkdirs()
        }

        val buffer = File(pathForID, "main.json").sink().buffer()

        if (BuildConfig.DEBUG) {
            val of = com.squareup.moshi.JsonWriter.of(buffer)
            of.indent = "  "
            jsonAdapter.toJson(of, pass as PassImpl)
            buffer.close()
            of.close()
        } else {
            jsonAdapter.toJson(buffer, pass as PassImpl)
            buffer.close()
        }

        passMap[pass.id] = pass
    }

    private fun readPass(id: String): Pass? {
        val pathForID = getPathForID(id)
        val language = context.resources.configuration.locale.language

        if (!pathForID.exists() || !pathForID.isDirectory) {
            return null
        }

        val file = File(pathForID, "main.json")
        var result: Pass? = null
        var dirty = true
        if (file.exists()) {
            val jsonAdapter = moshi.adapter(PassImpl::class.java)
            dirty = false
            try {
                result = jsonAdapter.fromJson(file.source().buffer())
            } catch (ignored: JsonDataException) {
                tracker.trackException("invalid main.json", false)
            }
        }

        if (result == null && File(pathForID, "data.json").exists()) {
            result = PassReader.read(pathForID)
            File(pathForID, "data.json").delete()
        }

        if (result == null && File(pathForID, "pass.json").exists()) {
            result = AppleStylePassReader.read(pathForID, language, context, tracker)
        }

        if (result != null) {
            if (dirty) {
                save(result)
            }
            passMap[id] = result
            notifyChange()
        }

        return result
    }

    override fun getPassbookForId(id: String): Pass? {
        return passMap[id] ?: readPass(id)
    }

    override fun deletePassWithId(id: String): Boolean {
        val result = getPathForID(id).deleteRecursively()
        if (result) {
            passMap.remove(id)
            classifier.removePass(id)
            notifyChange()
        }
        return result
    }

    override fun getPathForID(id: String): File {
        return File(path, id)
    }

    override fun notifyChange() {
        GlobalScope.launch {
            updateChannel.send(PassStoreUpdateEvent)
        }
    }

    override fun syncPassStoreWithClassifier(defaultTopic: String) {
        val keysToRemove = classifier.topicByIdMap.keys.filter { getPassbookForId(it) == null }

        for (key in keysToRemove) {
            classifier.topicByIdMap.remove(key)
        }

        val allPasses = path.listFiles()
        allPasses?.forEach {
            classifier.getTopic(it.name, defaultTopic)
        }
    }
}
